### PR TITLE
Extract updater to Akeneo\Pim\Structure

### PR DESCRIPTION
**Pull request by arnolanglade**
**Description (for Contributor and Core Developer)**

After moving all models related to the product structure, we have been moved the updaters in this PR. The attribute updater was decorated in EE to manage specific properties. The CE updaters has been improved, now, it better manage the attribute properties so we don't need decorate it.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | yes
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 1]
```diff
@@ -79,16 +79,16 @@ class FamilyController
...
-     * @param SaverInterface             $saver
-     * @param RemoverInterface           $remover
-     * @param ValidatorInterface         $validator
-     * @param SecurityFacade             $securityFacade
-     * @param FamilyFactory              $familyFactory
-     * @param NormalizerInterface          $constraintViolationNormalizer
+     * @param FamilyRepositoryInterface                             $familyRepository
+     * @param NormalizerInterface                                   $normalizer
+     * @param FamilySearchableRepository                            $familySearchableRepo
+     * @param \Akeneo\Pim\Structure\Component\Updater\FamilyUpdater $updater
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: import it

File: src/Akeneo/Pim/Structure/Component/spec/Updater/FamilyUpdaterSpec.php [hunk 2]
```diff
@@ -641,7 +641,7 @@ function it_throws_an_exception_when_an_attribute_in_attribute_requirements_is_n
                 InvalidPropertyTypeException::validArrayStructureExpected(
                     'attribute_requirements',
                     'one of the attributes in the channel "tablet" is not a scalar',
-                    'Pim\Component\Catalog\Updater\FamilyUpdater',
+                    'Akeneo\Pim\Structure\Component\Updater\FamilyUpdater',
```
Comment (skeleton) [inline src/Akeneo/Pim/Structure/Component/spec/Updater/FamilyUpdaterSpec.php]: I would prefer `FamilyUpdater::class` (for all specs)

Comment (arnolanglade) [reply src/Akeneo/Pim/Structure/Component/spec/Updater/FamilyUpdaterSpec.php]: Me too ;) but we won't rework all specs. The team which will be in charge of each will do that.

File: src/Akeneo/Pim/Structure/Component/Updater/AttributeUpdater.php [hunk 3]
```diff
@@ -181,11 +186,12 @@ protected function setData(AttributeInterface $attribute, $field, $data)
             case 'allowed_extensions':
                 $attribute->setAllowedExtensions(implode(',', $data));
                 break;
-            case 'auto_option_sorting':
-                $attribute->setProperty('auto_option_sorting', $data);
-                break;
             default:
-                $this->setValue($attribute, $field, $data);
+                if (in_array($field, $this->properties)) {
```
Comment (ahocquard) [inline src/Akeneo/Pim/Structure/Component/Updater/AttributeUpdater.php]: less readable for almost no gain.

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/updaters.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            - '@pim_localization.updater.translatable'
+
+    pim_catalog.updater.attribute:
+        class: 'Akeneo\Pim\Structure\Component\Updater\AttributeUpdater'
+        arguments:
+            - '@pim_catalog.repository.attribute_group'
+            - '@pim_catalog.repository.locale'
+            - '@pim_catalog.registry.attribute_type'
+            - '@pim_localization.updater.translatable'
+            - ['auto_option_sorting']
```
Comment (ahocquard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/updaters.yml]: I'm not a big fan to put it through DI. Any reason why?

File: UPGRADE.md [hunk 5]
```diff
@@ -121,3 +121,7 @@ find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Bundle\\CatalogBundle\\Ent
...
 find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Catalog\\Model\\VariantAttributeSet/Akeneo\\Pim\\Structure\\Component\\Model\\VariantAttributeSet/g'
 find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Catalog\\Model\\VariantAttributeSetInterface/Akeneo\\Pim\\Structure\\Component\\Model\\VariantAttributeSetInterface/g'
+find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Catalog\\Updater\\AttributeUpdater/Akeneo\\Pim\\Structure\\Component\\Updater\\AttributeUpdater/g'
+find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Catalog\\Updater\\AttributeOptionUpdater/Akeneo\\Pim\\Structure\\Component\\Updater\\AttributeOptionUpdater/g'
+find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Catalog\\Updater\\FamilyUpdater/Akeneo\\Pim\\Structure\\Component\\Updater\\FamilyUpdater/g'
+find ./src/ -type f -print0 | xargs -0 sed -i 's/Akeneo\\Tool\\Component\\Api\\Updater\\FamilyVariantUpdater/Akeneo\\Pim\\Structure\\Component\\Updater\\ExternalApi\\FamilyVariantUpdater/g'
```
Comment (ahocquard) [inline UPGRADE.md]: miss family variant updater (but not external)

File: src/Akeneo/Pim/Structure/Component/Updater/AttributeUpdater.php (hunk 3 above)
Comment (arnolanglade) [reply src/Akeneo/Pim/Structure/Component/Updater/AttributeUpdater.php]: I don't understand!

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/updaters.yml (hunk 4 above)
Comment (arnolanglade) [reply src/Akeneo/Pim/Structure/Bundle/Resources/config/updaters.yml]: yes it was a quick fix, to remove the EE updater.

**Pull request merged**